src/main/java/net/minecraftforge/event/terraingen/ChunkGeneratorEvent.java

```diff
@@ -106,4 +114,195 @@ public InitNoiseField(IChunkGenerator chunkProvider, double[] noisefield, int po
         public int getSizeY() { return sizeY; }
         public int getSizeZ() { return sizeZ; }
     }
+
+    /**
+     * This event is fired during chunk generation as well as structure regeneration before any of the vanilla structure gens are executed.
+     * If fired during recreateStructures the chunkprimer will be null
+     * This event is fired regardless of mapFeaturesEnabled.
+     */
+    public static class Generate extends ChunkGeneratorEvent
+    {
+        private final World world;
+        private final @Nullable ChunkPrimer primer;
+        private final int x, z;
+
+        public Generate(IChunkGenerator gen, World world, int x, int z, @Nullable ChunkPrimer primer)
+        {
+            super(gen);
+            this.world = world;
+            this.x = x;
+            this.z = z;
+            this.primer = primer;
+        }
+
+        public World getWorld()
+        {
+            return world;
+        }
+
+        public @Nullable ChunkPrimer getPrimer()
+        {
+            return primer;
+        }
+
+        public int getX()
+        {
+            return x;
+        }
+
+        public int getZ()
+        {
+            return z;
+        }
+    }
+
+    /**
+     * This event is fired when the chunk generator is asked for a list of possible creatures for a specific location.
+     * (Vanilla checks if there is a structure which wants to spawn special creatures at the location)
+     * <p>
+     * To return your own spawn entries, add them to the list and cancel this event.
+     */
+    @Cancelable
+    public static class GetPossibleCreatures extends ChunkGeneratorEvent
+    {
+        private final EnumCreatureType creatureType;
+        private final List<Biome.SpawnListEntry> possibleCreatures;
+        private final World world;
+        private final BlockPos pos;
+
+        public GetPossibleCreatures(IChunkGenerator gen, World world, EnumCreatureType creatureType, BlockPos pos)
+        {
+            super(gen);
+            this.creatureType = creatureType;
+            this.pos = pos;
+            this.world = world;
+            possibleCreatures = Lists.newArrayList();
+        }
+
+        public World getWorld()
+        {
+            return world;
+        }
+
+        public EnumCreatureType getCreatureType()
+        {
+            return creatureType;
+        }
+
+        public BlockPos getPos()
+        {
+            return pos;
+        }
+
+        public List<Biome.SpawnListEntry> getPossibleCreatures()
+        {
+            return possibleCreatures;
+        }
+
+    }
+
+    /**
+     * This event is fired if the chunk generator is asked if there is a structure at a specific location.
+     * <p>
+     * To return your own value set isInStructure and cancel this event.
```

Comment:
same here

Reply:
This requires some kind of cancellation as well to determine if the vanilla code should be used or not.
However, since this is a "three option" kind of event (no-structure, structure, vanilla check), I could use the @HasResult here (Allow,Deny,Default), but I think it does not make a big difference here.